HHS/Head-Start-TTADP
# [TTAHUB-1990] Goal similarity API

## Description of change

This is almost entirely @kcirtapfromspace's code from his NLP PR: https://github.com/HHS/Head-Start-TTADP/pull/1561

There are enough changes made that I think this warranted a different branch & PR, rather than compounding the existing branch & PR with commits that fundamentally change it.

#### What didn't change:

- The model being used.
- The similarity calculations for all goals given a recipient ID, and how this calculation is made.
- Most of the CircleCI and CF configuration for this project.
- Most of the docker-compose configuration.

#### What did change/was added:

- The name, from data_science to similarity.
- Configurable `alpha` for computation requests.
- Add two other types of similarity calculations.
  - Given a recipient ID and a goal text, find all other goals for this recipient that are within `alpha` similarity, and return them. This is fast enough that we can probably use this as a typeahead source.
  - Given two strings, just return their similarity score. This is a general-purpose calculation that I'm not sure we have use for at the moment, but may be useful in the future.
- Authentication is now just an API key passed as the `X-API-KEY` header.
- The OpenAPI schema is no longer automatically generated. The API has only one endpoint, so I just wrote it out manually.

#### Thoughts:

Should we cache similarity scores in a table with, e.g., three columns: (goal1, goal2, score) that are invalidated in a hook when the goal name is changed/deleted?

## How to test

- `docker-compose down`
- Check out this branch
- `docker-compose up`

At this point, assuming no issues, the similarity API should be alive and ready at `localhost:9100`. There is a file included at `similarity_api/src/test.http` that covers all possible ways to call the `/compute` endpoint. The pre-written requests reference goal names that exist in the transformed SQL dump on keybase.

I use [this extension](https://marketplace.visualstudio.com/items?itemName=humao.rest-client) for VSCode to run these requests. It puts a little "Send Request" action above each request block:

![Screenshot 2023-09-26 at 10 15 51 AM](https://github.com/HHS/Head-Start-TTADP/assets/17074357/add96dff-8f81-4aa1-8eaf-cdff6bf0feee)

Or, you can use this file as a reference to craft requests using something like curl or postman.


## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1990


## Checklists

### Every PR



### Production Deploy


### After merge/deploy
